src/rust/engine/src/rule_graph.rs

```diff
@@ -602,6 +690,54 @@ impl RuleGraph {
   }
 }
 
+impl fmt::Display for RuleGraph {
```

Comment:
This seems like an abuse of `Display`. I wouldn't accidentally want to drop this in a log message, for example.